src/qililab/automatic_calibration/calibration_node.py

```diff
@@ -0,0 +1,682 @@
+        """
+        # Remove .ipynb from end if it has:
+        shorted_path = original_path.split(".ipynb")[0]
+
+        # remove anything after the last "/":
+        folder_path_list = shorted_path.split("/")
+        if isinstance(self.qubit_index, int):
+            qubit_str = f"_q{str(self.qubit_index)}"
+        elif isinstance(self.qubit_index, list):
+            qubit_str = "_"
+            for q in self.qubit_index:
+                qubit_str += f"q{str(q)}"
+        else:
+            qubit_str = ""
+
+        name = folder_path_list.pop() + qubit_str
+        folder_path = "/".join(folder_path_list)
+
+        return name, folder_path
+
+    def get_last_calibrated_timestamp(self) -> float | None:
+        """Gets the last executed timestamp if there exist any previous execution of the same notebook.
+
+        Returns:
+            float: The last execution timestamp if exists, None otherwise.
+        """
+        # get all elements in that folder starting with self.node_id+“_” and get the last modified one file name, convert string into datetime
+        last_modified_file_name = self._find_last_executed_calibration()
+        return (
+            os.path.getmtime(f"{self.nb_folder}/{last_modified_file_name}")
+            if last_modified_file_name is not None
+            else None
+        )
+
+    def get_last_calibrated_output_parameters(self) -> dict | None:
+        """Gets the last output of the previous calibration execution of the same notebook if there exist any.
+
+        Returns:
+            dict: The last calibration execution output if exists, None otherwise.
+        """
+        last_modified_file_name = self._find_last_executed_calibration()
+        return (
+            self._parse_output_from_execution_file(last_modified_file_name)
+            if last_modified_file_name is not None
+            else None
+        )
+
+    def _find_last_executed_calibration(self) -> str | None:
+        """Returns the filename of the last calibration execution with the same node_id.
+
+        Returns:
+            str: Name of the file.
+        """
+        # Filtering by node_id and calibrations
+        entries = os.scandir(self.nb_folder)
+        file_names = []
+        for val in entries:
+            if val.is_file():
+                same_node_fnames = val.name.split(f"{self.node_id}_")
+                if len(same_node_fnames) == 2 and len(val.name.split("_calibrated")) == 2:
+                    file_names.append(val.name)
+
+        # Get the last created file, most recent one
+        last_modified_file_time, last_modified_file_name = -1.0, ""
+        for fname in file_names:
+            ftime = os.path.getctime(f"{self.nb_folder}/{fname}")
+            if ftime > last_modified_file_time:
+                last_modified_file_time, last_modified_file_name = ftime, fname
+
+        return last_modified_file_name if last_modified_file_time != -1.0 else None
```

Comment:
Try this, I think it does the same thing:


```suggestion
        file_names = [entry.name for entry in os.scandir(self.nb_folder) if entry.is_file() and f"{self.node_id}_" in entry.name and "_calibrated" in entry.name]
        return max(file_names, key=lambda fname: os.path.getctime(os.path.join(self.nb_folder, fname)), default=None)
```

Reply:
@IsaacLA12 I leave this one to you!

Reply:
Yeah for the second `for` loop, that looks cleaner and nice!!! For the first `for` loop I had thought using `in` to serch the substring in the file_name but had a problem, I thought that maybe someone could name a notebook "_calibrated" and that will pass (actually my implementation is not checking it properly cuz I found a typo hehe) But I dont knw if we should think that much on that because any user could find a way to modify a filename to "trick" the method I supose 🤔

Reply:
Nevertheless, I still changed the loop for a comprehension list!

Reply:
Yess in the end we cannot and we should not be god and read minds, I think what you did now is more than sufficient, thanks Isaac!

Reply:
Closing this, since solved!